src/native/corehost/corehost.cpp

```diff
@@ -93,6 +100,44 @@ void need_newer_framework_error()
 
 #if defined(CURHOST_EXE)
 
+int try_run_as_arch(const pal::char_t *path, int rc, const int argc, const pal::char_t* argv[])
+{
+    assert((rc == StatusCode::CorehostMustRunAsArchArm64) || (rc == StatusCode::CorehostMustRunAsArchX64));
+#if !defined(FEATURE_APPHOST) && defined(TARGET_OSX)
+
+    trace::verbose(_X("Restarting process in arch %s."), (rc == CorehostMustRunAsArchArm64) ? _X("Arm64") : _X("x64"));
+
+    posix_spawnattr_t attr;
+    int result = 0;
+    size_t size;
+
+    int curproc_arch_affinity = (rc == CorehostMustRunAsArchArm64) ? 0 : 1;
+    size = sizeof(curproc_arch_affinity);
+    sysctlbyname("kern.curproc_arch_affinity", NULL, NULL, &curproc_arch_affinity, size);
+    // kern.curproc_arch_affinity set ignore errors (not available on x64 platform)
+
+    result = posix_spawnattr_init(&attr);
+
+    if (result == 0)
+        result = posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETEXEC);
+
+    if (result == 0)
+    {
+        cpu_type_t cpu_pref = (rc == CorehostMustRunAsArchArm64) ? CPU_TYPE_ARM64 : CPU_TYPE_X86_64;
+        result = posix_spawnattr_setbinpref_np(&attr, 1, &cpu_pref, &size);
```

Comment:
Is `posix_spawnattr_setarchpref_np` supported on all macOS version that we support?

Reply:
From the man pages.

```
HISTORY
     The posix_spawnattr_setbinpref_np() and posix_spawnattr_getbinpref_np() function calls 
     appeared in Mac OS X.
```

```
HISTORY
     The posix_spawnattr_setarchpref_np() and posix_spawnattr_getarchpref_np() function calls 
     appeared in macOS 10.16
```

I chose to use the older interface `posix_spawnattr_setbinpref_np` because it didn't require a `cpu_subtype_t`.  It is also apparently more available.

Reply:
We are currently compiling x64 for macos 10.13.  So no not `posix_spawnattr_setarchpref_np`

Reply:
Looks like 10.13 is unsupported, but 10.14 is still actively supported.